Add common Azure IoT Hub topic parser

jtguggedal opened the pull request:
This PR adds the following:
- Common Azure IoT Hub topic parser
- Suppoort for property bags, both in received messages and when sending messages
- Tests for the topic parser

coderbyheart commented on subsys/net/lib/azure_iot_hub/src/azure_iot_hub.c:
```diff
@@ -32,17 +33,23 @@ LOG_MODULE_REGISTER(azure_iot_hub, CONFIG_AZURE_IOT_HUB_LOG_LEVEL);
+#define TOPIC_TWIN_REPORT	"$iothub/twin/PATCH/properties/reported/?$rid=%d"
+#define TOPIC_EVENTS		"devices/%s/messages/events/%s"
+#define TOPIC_TWIN_REQUEST	"$iothub/twin/GET/?$rid=%d"
+#define TOPIC_DIRECT_METHOD_RES	"$iothub/methods/res/%d/?$rid=%d"
```
`request ID` is a string, not just an integer.

jtguggedal replied:
Is it? I wasn't aware. My (limited) observations until now has been that it's a counter starting from 1 on the first method invocation after a connection is established. I can change to string if it's expected that it can send characters in the ID.

coderbyheart replied:
[The documentation](https://docs.microsoft.com/en-us/azure/iot-hub/iot-hub-devguide-direct-methods) gives the impression that one could pass anything in the request ID:

> Direct method invocations on a device are HTTPS calls that are made up of the following items:
> ...
> Headers that contain the authorization, request ID, content type, and content encoding.

However it is not documented as a parameter in https://docs.microsoft.com/en-us/rest/api/iothub/service/devices/invokemethod

So I am unsure if it's possible and I need to investigate.

coderbyheart replied:
When you invoke a direct method from the UI, this is the topic string it arrives on: `$iothub/methods/POST/test/?$rid=1`, so here `rid` is an integer.

When I invoke a direct method with a different name from CLI the rid increases.

```
az iot hub generate-sas-token -n bifravstprodIotHub --du 3600

curl -X POST \
  https://bifravstprodIotHub.azure-devices.net/twins/e66b9f4a-f001-4dcc-8185-4665bffc59fe/methods?api-version=2018-06-30 \
  -H 'Authorization: <sas>' \
  -H 'Content-Type: application/json' \
  -d '{
    "methodName": "reboot",
    "responseTimeoutInSeconds": 200,
    "payload": {
        "input1": "someInput",
        "input2": "anotherInput"
    }
}'
```

```
< $iothub/methods/POST/reboot/?$rid=3
< {"input1":"someInput","input2":"anotherInput"}
```

So, I conclude that `rid` is indeed an integer, and managed by the IoT hub itself.

coderbyheart commented on tests/subsys/net/lib/azure_iot_hub/src/main.c:
```diff
@@ -0,0 +1,234 @@
+	zassert_true(strcmp(topic_data.prop_bag[1].value, "3") == 0, NULL);
+}
+
+static void test_topic_parse_prop_bag_overload(void)
```
The test key/value data for property bags are pure ascii, however they are RFC 2396-encoded values, so this should also work as key and/or value: `det%20skal%20ogs%C3%A5%20funker%20%F0%9F%98%87` and parsed into `det skal også funker 😇` (however I don't know if Zephyr supports unicode strings).

Same goes for adding values.

Please extend the tests with some more complicated examples.

jtguggedal replied:
Yeah, that's why the PR is a draft. I haven't figured out a good way to solve this. It adds quite a bit of complexity to the implementation.

jtguggedal commented:
I've updated the PR and added the following:
- Tests for more complex input, though still only handling ASCII (not documented yet)
- Detecting property bag section in topic both with and without `?` to mark its starting point
- Forwarding percentage-encoded content in keys and values without attempting to decode it (not documented yet)

I'm still on the fence about the heap usage to allocate buffer for encoded property bag strings. I'll probably get rid of it, as the topic buffer that eventually goes into is fixed-length anyway.

rlubos reviewed (approved):
Looks good, just some (very minor) nits.

jtguggedal commented:
@rlubos I've addressed your comments.

I also added missing propagation of property bag content in case of devicebound messages (https://github.com/nrfconnect/sdk-nrf/compare/3521b6d3bf5060dbd9d43da86fffe65475e8830d..c0090dc0e3c3e2b3efedd0fc88bc62fb66f43ab9)

simensrostad commented on subsys/net/lib/azure_iot_hub/src/azure_iot_hub_topic.c:
```diff
@@ -0,0 +1,335 @@
+		LOG_DBG("Dynamic value: %s", log_strdup(data->name));
+
+		if ((data->type == TOPIC_TYPE_TWIN_UPDATE_RESULT) ||
+		(data->type == TOPIC_TYPE_DPS_REG_RESULT)) {
```
Alignment

simensrostad commented on tests/subsys/net/lib/azure_iot_hub/src/main.c:
```diff
@@ -0,0 +1,297 @@
+	zassert_not_null(prop_bag_str, NULL);
+	zassert_equal(strcmp(prop_bag_str, "?key3&key2=&key1=value1"), 0,
+		      "Incorrect property bag string");
+	k_free(prop_bag_str);
```
Maybe add a static inline function in ```azure_iot_hub_topic.h``` that implicitly frees ```property_bag_str``` type pointers? Makes it more intuitive looking at the API.

jtguggedal replied:
Yep, that makes it more clear what happens. Added

simensrostad commented on tests/subsys/net/lib/azure_iot_hub/src/main.c:
```diff
@@ -0,0 +1,297 @@
+		      "Incorrect property bag count");
+	zassert_true(strcmp(topic_data.prop_bag[0].key, "%24.mid") == 0, NULL);
+	zassert_true(strcmp(topic_data.prop_bag[0].value,
+		     "456132-235-a2fd-8458-56432854d") == 0, NULL);
```
Alignment?

simensrostad commented on tests/subsys/net/lib/azure_iot_hub/testcase.yaml:
```diff
@@ -0,0 +1,4 @@
+tests:
+  net.lib.azure_iot_hub:
+    platform_whitelist: nrf9160dk_nrf9160 qemu_x86 native_posix
```
This is now changed to ```platform_allow```

simensrostad reviewed (commented):
Looks good. Left a couple of comments to consider.

jtguggedal commented on subsys/net/lib/azure_iot_hub/src/azure_iot_hub_dps.c:
```diff
@@ -367,36 +357,67 @@ static void dps_parse_reg_update(char *topic, size_t topic_len, char *payload,
-	/* Check that request ID matches the expected value */
-	(void)strtok(NULL, "="); /* Pointer at start of "$rid" */
-	ptr = strtok(NULL, "&"); /* Pointer at start of "{request_id}" */
+	/* Usually the retry-after property comes second. */
```
It should be handled by the current logic, I think.

jtguggedal commented on subsys/net/lib/azure_iot_hub/src/azure_iot_hub.c:
```diff
@@ -187,121 +194,23 @@ static int topic_subscribe(void)
-	memcpy(rid_buf, ptr, max_ptr - ptr + 1);
-	rid_buf[max_ptr - ptr + 1] = '\0';
-	evt.data.method.rid = atoi(rid_buf);
+	evt.data.method.rid = atoi(topic->prop_bag[0].value);
```
Added checks for property bag count and direct method name

jtguggedal commented:
Rebased after upmerge and addressed comments from @simensrostad

simensrostad commented on tests/subsys/net/lib/azure_iot_hub/src/main.c:
```diff
@@ -0,0 +1,297 @@
+	zassert_not_null(prop_bag_str, NULL);
+	zassert_equal(strcmp(prop_bag_str, "?key1=value1&key2=&key3"), 0,
+		      "Incorrect property bag string");
+	k_free(prop_bag_str);
```
```azure_iot_hub_prop_bag_free()``` can be used here.

simensrostad commented on tests/subsys/net/lib/azure_iot_hub/src/main.c:
```diff
@@ -0,0 +1,297 @@
+	zassert_not_null(prop_bag_str, NULL);
+	zassert_equal(strcmp(prop_bag_str, "?key3&key2=&key1=value1"), 0,
+		      "Incorrect property bag string");
+	k_free(prop_bag_str);
```
and here.

coderbyheart commented on subsys/net/lib/azure_fota/azure_fota.c:
```diff
@@ -210,6 +210,9 @@ static bool parse_reported_status(const char *msg)
 	if (fw_obj == NULL) {
 		LOG_DBG("No 'firmware' object found");
+
+		report_needed = true;
```
Unrelated change?

jtguggedal replied:
It's intentional, in a separate commit: https://github.com/nrfconnect/sdk-nrf/pull/2962/commits/d831092bde9335e9798bb5a9d71b69cd4592a0f4
Kept it in the PR to not increase the CI load git such a small change.

coderbyheart commented on subsys/net/lib/azure_iot_hub/Kconfig:
```diff
@@ -71,6 +71,34 @@ config AZURE_IOT_HUB_AUTO_DEVICE_TWIN_REQUEST
+	help
+	  Sets the maximum length of a topic. If a user plans to make extensive
+	  use of property bags, it should be considered to increase this value
+	  to fit the topic in the allocated buffers.
```
```suggestion
	  to fit the topic into the allocated buffers.
```

coderbyheart commented on subsys/net/lib/azure_iot_hub/include/azure_iot_hub_topic.h:
```diff
@@ -0,0 +1,118 @@
+	const char *topic;
+	/* Length of topic buffer. */
+	size_t topic_len;
+	/* Each topic may contais one dynamic topic level that carries
```
```suggestion
	/* Each topic may contain one dynamic topic level that carries
```

coderbyheart commented on subsys/net/lib/azure_iot_hub/include/azure_iot_hub_topic.h:
```diff
@@ -0,0 +1,118 @@
+	};
+	/* Array of property bag elements. */
+	struct topic_parser_prop_bag prop_bag[TOPIC_PROP_BAG_COUNT];
+	/* Number of property bag elements  in the above array. */
```
```suggestion
	/* Number of property bag elements in the above array. */
```

simensrostad reviewed (approved):
LGMT

coderbyheart commented on subsys/net/lib/azure_iot_hub/include/azure_iot_hub_topic.h:
```diff
@@ -0,0 +1,118 @@
+/* @brief Parse topic.
+ *
+ * @param data Pointer to topic data structure. The structure must
+ *	       be initialized with topic buffer pointer and topic
```
```suggestion
 *	       be initialized with a topic buffer pointer and topic
```

coderbyheart commented on subsys/net/lib/azure_iot_hub/src/azure_iot_hub.c:
```diff
@@ -486,14 +419,21 @@ static void on_publish(struct mqtt_client *const client,
-	} else {
-		azure_iot_hub_notify_event(&evt);
+		return;
+	case TOPIC_TYPE_UNKNOWN:
+		LOG_WRN("Data received on unknown topic");
+		break;
```
The topic is **unexpected**, because the name of topic is known.

coderbyheart commented on subsys/net/lib/azure_iot_hub/src/azure_iot_hub_topic.c:
```diff
@@ -0,0 +1,341 @@
+ *
+ * @retval the length of parsed property bag on success.
+ * @retval 0 if there was no property bag the provided buffer.
+ * @retval -ENOMEM nif the key or value is too long for the buffer set by
```
```suggestion
 * @retval -ENOMEM if the key or value is too long for the buffer set by
```

coderbyheart commented on subsys/net/lib/azure_iot_hub/src/azure_iot_hub_topic.c:
```diff
@@ -0,0 +1,341 @@
+enum topic_type topic_type_get(const char *buf, const size_t len)
+{
+	if (buf == NULL || len == 0) {
+		return TOPIC_TYPE_UNKNOWN;
```
`TOPIC_TYPE_EMPTY` would be better.

coderbyheart commented on subsys/net/lib/azure_iot_hub/src/azure_iot_hub_topic.c:
```diff
@@ -0,0 +1,341 @@
+		}
+	}
+
+	return TOPIC_TYPE_UNKNOWN;
```
`TOPIC_TYPE_UNEXPECTED` would be better here.

coderbyheart commented on subsys/net/lib/azure_iot_hub/src/azure_iot_hub_topic.c:
```diff
@@ -0,0 +1,341 @@
+			stop_ptr += sizeof(TOPIC_DEVICE_BOUND_SUFFIX) - 1;
+		}
+
+		/* Move the start pointer to the end of the method name plus the
+		* subsequent '/'. start_ptr will now be at the start of the
+		* propery bags, if there are any.
+		*/
```
```suggestion
		/* Move the start pointer to the end of the method name plus the
		 * subsequent '/'. start_ptr will now be at the start of the
		 * property bags, if there are any.
		 */
```

coderbyheart commented on subsys/net/lib/azure_iot_hub/include/azure_iot_hub_topic.h:
```diff
@@ -0,0 +1,118 @@
+	size_t prop_bag_count;
+};
+
+/* @brief Free an allocated propery bag buffer.
```
```suggestion
/* @brief Free an allocated property bag buffer.
```

jtguggedal commented:
@coderbyheart Your comments have been addressed. Typos are fixed and a couple of topic type enums added as suggested
